backends/hip/ceed-hip-qfunction.c

```diff
@@ -175,15 +114,20 @@ int CeedQFunctionCreate_Hip(CeedQFunction qf) {
   CeedChkBackend(ierr);
 
   // Read source
-  char *source;
-  ierr = CeedQFunctionGetSourcePath(qf, &source); CeedChkBackend(ierr);
-  const char *funname = strrchr(source, ':') + 1;
-  data->qFunctionName = (char *)funname;
-  const int filenamelen = funname - source;
-  char filename[filenamelen];
-  memcpy(filename, source, filenamelen - 1);
-  filename[filenamelen - 1] = '\0';
-  ierr = CeedHipLoadQFunction(qf, filename); CeedChkBackend(ierr);
+  char *source_path;
+  ierr = CeedQFunctionGetSourcePath(qf, &source_path); CeedChkBackend(ierr);
+  // Empty source path indicates user must supply Q-Function
+  if (source_path[0] != '\0') {
+    ierr = CeedQFunctionGetKernelName(qf, &data->qFunctionName);
+    CeedChkBackend(ierr);
+    ierr = CeedLoadSourceToBuffer(ceed, source_path, &data->qFunctionSource);
+    CeedChkBackend(ierr);
+  } else {
+    data->module = NULL;
+    data->qFunctionName = "";
+    data->qFunctionSource = "";
+    data->qFunction = NULL;
+  }
```

Comment:
Actually this whole pattern seems to repeat, so the whole pattern should probably be a function.